sematic/external_resource.py

```diff
@@ -0,0 +1,331 @@
+    ACTIVE:
+        The resource is allocated and ready for use.
+    DEACTIVATING:
+        The resource is being deactivated, and likely is no longer usable.
+    DEACTIVATED:
+        The resource has been deactivated and may not even exist anymore.
+        It is not usable.
+    """
+
+    CREATED = "CREATED"
+    ACTIVATING = "ACTIVATING"
+    ACTIVE = "ACTIVE"
+    DEACTIVATING = "DEACTIVATING"
+    DEACTIVATED = "DEACTIVATED"
+
+    def is_allowed_transition(self, other_state: "ResourceState") -> bool:
+        """True if going from the current state to the other is allowed, otherwise False.
+
+        Parameters
+        ----------
+        other_state:
+            The state being transitioned to
+
+        Returns
+        -------
+        True if going from the current state to the other is allowed, otherwise False.
+        """
+        return other_state in _ALLOWED_TRANSITIONS[self]
+
+    def is_terminal(self) -> bool:
+        """True if there are no states that can follow this one, False otherwise."""
+        return len(_ALLOWED_TRANSITIONS[self]) == 0
+
+
+_ALLOWED_TRANSITIONS = {
+    None: {ResourceState.CREATED},
+    ResourceState.CREATED: {
+        # Created -> Activating: normal progression for activation
+        ResourceState.ACTIVATING,
+        # Created -> Deactivated: after creating the resource,
+        # but before starting to activate it, it was decided
+        # to not activate the resource after all. Since no activation
+        # was performed, there is no need to do anything for deactivation
+        # so we can skip DEACTIVATING.
+        ResourceState.DEACTIVATED,
+    },
+    ResourceState.ACTIVATING: {
+        # Activating -> Active: normal progression for successful activation
+        ResourceState.ACTIVE,
+        # Activating -> Deactivating: activation failed, immediate deactivation
+        ResourceState.DEACTIVATING,
+    },
+    ResourceState.ACTIVE: {
+        # Active -> Deactivating:
+        #   - possibly normal termination, due to no longer being needed.
+        #   - possibly an error with the resource. Status message should have more info.
+        ResourceState.DEACTIVATING,
+    },
+    ResourceState.DEACTIVATING: {
+        # Deactivating -> Deactivated: normal progression for deactivation
+        ResourceState.DEACTIVATED,
+    },
+    ResourceState.DEACTIVATED: {},
+}
+
+
+@dataclass(frozen=True)
+class ResourceStatus:
+    """The status of the resource.
+
+    Attributes
+    ----------
+    state:
+        The discrete state that the resource is in
+    message:
+        A human-readable message about why the resource entered this state
+    last_update_epoch_time:
+        The last time this status was checked against the actual external resource
+    """
+
+    state: ResourceState
+    message: str
+    last_update_epoch_time: int = field(default_factory=lambda: int(time.time()))
+
+
+T = TypeVar("T")
+
+
+@dataclass(frozen=True)
+class ExternalResource:
```

Comment:
Thoughts on calling this `Resource`? Too generic/abstract?

Reply:
yes, especially considering when we already have `ResourceRequirements` to refer to the compute resources to execute the body of a func's code